src/LCD_Fonts.h

```diff
@@ -0,0 +1,43 @@
+/**
+ * A library to control the monochrome LCD displays found on Nokia 5110.
+ *
+ * It is a derivative of Jim Lindblom's work, from Sparkfun. The goal
+ * was to decrease its size so it can be ran on ATTiny microcontrollers
+ * that have limited RAM size. Therefore, the main difference is that a
+ * framebuffer, which would consume all dynamic memory, is not used. Get the
+ * original library [here](https://github.com/sparkfun/GraphicLCD_Nokia_5110).
+ *
+ * On how to use this library, please refer to the `examples/` folder.`
+ */
+#pragma once
+#include <Arduino.h>
+#include "Nokia_LCD_Fonts.h"
+class LcdFont
+{
+public:
+    LcdFont(const unsigned char* fontTable, 
```

Comment:
Is this correct :arrow_right: `const unsigned char* fontTable` ?
We want to pass a 2-dimensional array, but the constructor accepts a single-dimensional one.
Of course we can think of 2-d arrays as 1-d  ones, but since we already represent them in 2-D, why change that?

Have a look at the suggestion I previously sent. That one takes a 2-d array, where the "second" dimension is known at compile time (i.e. `Nokia_LCD_Fonts::kColumns_per_character`) and the class holds a pointer to it (i.e. `const unsigned char (*kFontTable)[Nokia_LCD_Fonts::kColumns_per_character]`).
Is there a reason why the suggestion wouldn't work? On the up side, the user doesn't need to cast the 2-d array to a `const unsigned char*` every time they want to pass it to the constructor.

You can try just the particular class out here: https://onlinegdb.com/3VLNdbM8U

Reply:
Okay, let me try to explain, maybe it could be a lack of understanding on my part.

There are two premises that are guiding me, it's better to make them clear here:
- Custom fonts are not inside the library, they are in the project that is using it. So I need a common class for any font type.
- As you suggested, I'm avoiding using inheritance, just a simple class where the objects instantiated are fonts of different sizes.

In your example, NokiaLcdFont works great but only for 5x8 fonts
`NokiaLcdFont f{ fontTable, hSpace, 1, 0 };`

I have a font called SmallFont with 3x8 and another, bold with 7x8. I could create the classes SmallLcdFont and BoldLcdFont respectively, like you did with NokiaLcdFont. Bbut I wouldn't be able to pass them as a parameter to the library, since they are different classes and don't have a common parent.

`Nokia_LCD::setFont(const ??? *font)` <- ??? what type here?

So I created an LcdFont class that receives fonts of any size and that's why the `const unsigned char* fontTable` was used. If you have a better solution, please let me know. I'm not an C++ expert and I maybe there is something that solves the problem gracefully.

Reply:
I updated the gist https://gist.github.com/jnthas/81e5021417fed9aee8cdcb3a35956a45.
Think that the class example.cpp is outside of the library and just importing it.

Reply:
> In your example, NokiaLcdFont works great but only for 5x8 fonts

OK, I see what you mean. The class I suggested indeed accepts `5xColumns` fonts only. Let me think of it and I will get back to you soon.

Reply:
OK, there are different ways that we can adopt to solve the problem:
1. **Inheritance** like you initially suggested :arrow_right: This isn't a bad way to do things to be honest, but I fear there may make the size of the executable larger which is often an issue on ATtiny microcontrollers. The next drawback was that a user would have to create their own class which may or not be easy.
2. **Templatize** the `LcdFont` class :arrow_right: This would allow us to support pretty much all shapes and sizes. The problem is that we would need to do a large refactoring and turn the other classes into templates too. The change would be backward incompatible. 
3. Let the user bother with the array. This is similar to the first solution where we let the user define their own class and methods with the difference that the user wouldn't need to create their own class, but only a single function. I am currently leaning towards this, but let me show you how it would look like:

```cpp
using GetFontCallback = const unsigned char* (*)(char);

class LcdFont
{
public:
    LcdFont(GetFontCallback getFontCallback,
            uint8_t columnSize,
            const unsigned char* hSpace,
            uint8_t hSpaceSize,
            uint8_t rowSize = 8)
        : hSpace{ hSpace }
        , hSpaceSize{ hSpaceSize }
        , columnSize{ columnSize }
        , rowSize{ rowSize }
        , mGetFontCallback(getFontCallback)
    {
    }
    const unsigned char* getFont(char character) const
    {
        // We could have the offset here perhaps instead? I don't remember why we need it.
        return mGetFontCallback(character);
    }
    const unsigned char* const hSpace;
    const uint8_t hSpaceSize;
    const uint8_t columnSize;
    const uint8_t rowSize;

private:
    const GetFontCallback mGetFontCallback;
};
```

And a user would use it like this:

```cpp
const unsigned char fontTable[][3]{ { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 } };
unsigned char hSpace[]{ 0 };

void setup() {
    LcdFont f{ [](char c) { return fontTable[c - 0x20]; }, 3, hSpace, 1 };
}
```
Or if someone doesn't like lambdas, they alternatively could do:

```cpp
const unsigned char fontTable[][3]{ { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 } };

const unsigned char* getFont(char character)
{
    return fontTable[character - 0x20];
}


void setup() {
    LcdFont f{ getFont, 3, hSpace, 1 };
}
```

Let me know if you understand the solution and what you think of it. :+1:

Reply:
Ok, I got it. Let me change the code and see what happens. 
About the character offset, we should keep outside like you did, that way the users can have a full font table with 256 characters if they want.
And there is another thing that I noticed here now, we could remove rowSize from the constructor, it's optional there but it actually do not, the characters must be 8-bit size always.

Reply:
I like the idea very much, thanks for the suggestion. I just push the changes. I have two questions:
- I could instantiate the LcdFont object (nokiaFont) inside the `Nokia_LCD_Fonts` instead of the namespace of the main class, what do you think?
-  About the example folder, is it worth to create another example to demonstrate how to use a different font?

Reply:
> I could instantiate the LcdFont object (nokiaFont) inside the Nokia_LCD_Fonts instead of the namespace of the main class, what do you think?

Yeap, looking at your change you did it exactly like you should. :+1: 

> About the example folder, is it worth to create another example to demonstrate how to use a different font?

Absolutely! :100: